Premultiplied alpha on materials

bhouston opened the pull request:
Using premultiplied alpha on materials leads to better results for areas where there are specular highlights on transparency when using RGBA8 buffers.  This change should be perfectly backwards compatible, but allows for easy usage of premultiplied alpha when desired.

This was discussed here: https://github.com/mrdoob/three.js/issues/5810  Clara.io has been running for more than year with premultiplied alpha to great results.  I'd like to get this option into ThreeJS proper, at least as an option that people can use.

bhouston committed:
support pre-multipled alpha on materials.

bhouston committed:
minor fixes for premultiplied alpha option on materials.

mrdoob commented on src/renderers/webgl/WebGLProgram.js:
```diff
@@ -491,6 +491,8 @@ THREE.WebGLProgram = ( function () {
 				parameters.shadowMapEnabled ? '#define ' + shadowMapTypeDefine : '',
 				parameters.pointLightShadows > 0 ? '#define POINT_LIGHT_SHADOWS' : '',
 
+				parameters.premultipliedAlpha ? "#define PREMULTIPLIED_ALPHA" : '',
```
We need to modify `WebGLPrograms.js` too so `premultipliedAlpha` is added in `parameters`.

WestLangley commented:
As an alternative API, consider what I said in https://github.com/mrdoob/three.js/issues/5810#issuecomment-68198568:

>In fact, we could add a new blending option: THREE.PremultipyAlphaBlending, making the setting of the blending even easier.

That way, all you have to do in javascript is set:
```javascript
material.blending = THREE.PremultipyAlphaBlending;
```

The new property `material.premultipliedAlpha` is not required, and the user does not have to specify a custom blending mode.

The new `THREE.PremultipyAlphaBlending` blending option does this:
```javascript
_gl.blendEquation( _gl.FUNC_ADD );
_gl.blendFunc( _gl.ONE, _gl.ONE_MINUS_SRC_ALPHA );
```
In the shader, you would add this, instead:
```glsl
#ifdef PREMULTIPLIED_ALPHA_BLENDING

    gl_FragColor.rgb *= gl_FragColor.a;

#endif
```

bhouston committed:
forgotten file.

bhouston committed:
fix two bugs with premultiplied alpha implementation.

bhouston committed:
add premultiplied alpha example.

bhouston committed:
swithch from materia.premultipledAlpha to material.blend = THREE.PremultipiedAlphaBlending.

bhouston commented:
I've fixed a few bugs with the implementation, switch from the boolean material.premultipledAlpha to material.blend = THREE.PremultipledAlphaBlending and then I have added an example:

http://ci.threejs.org/api/pullrequests/8245/examples/webgl_materials_transparency.html

bhouston commented:
Here is an example of the different premultiplied alpha has - it is drastic:

![premutlipledalpha](https://cloud.githubusercontent.com/assets/588541/13399802/0a252788-ded3-11e5-8c19-47eaa46bebde.png)

bhouston committed:
spelling: PremultipiedAlphaBlending -> PremultipliedAlphaBlending

bhouston committed:
remove comment.

WestLangley commented on src/renderers/shaders/ShaderChunk/premultiplied_alpha_fragment.glsl:
```diff
@@ -0,0 +1,6 @@
+#ifdef PREMULTIPLIED_ALPHA
```
#ifdef PREMULTIPLIED_ALPHA_BLENDING

    gl_FragColor.rgb *= gl_FragColor.a;

#endif

We do not want this to be confused with `renderer.premultipliedAlpha`.

WestLangley commented on src/renderers/shaders/ShaderChunk/premultiplied_alpha_fragment.glsl:
```diff
@@ -0,0 +1,6 @@
+#ifdef PREMULTIPLIED_ALPHA
+
+  // Get get normal blending with premultipled, use with CustomBlending, OneFactor, OneMinusSrcAlphaFactor, AddEquation.
```
Comment not required, now.

WestLangley commented on src/renderers/shaders/ShaderLib/linedashed_frag.glsl:
```diff
@@ -30,5 +30,6 @@ void main() {
 	gl_FragColor = linearToOutputTexel( vec4( outgoingLight, diffuseColor.a ) );
 
 	#include <fog_fragment>
+	#include <premultiplied_alpha_fragment>
```
Consider `premultiplied_alpha_blending_fragment`.

bhouston commented:
@WestLangley thank you for the valuable feedback.  :)

You are right that I should make those changes.  One issue though is that I have already built another PR (https://github.com/mrdoob/three.js/pull/8250) on top of this one and it included other refactors.  Given that this feedback is about internals, and not about public interfaces, could we merge this and possibly the other PR (inline tonemapping) and then I can make these non-interface refactoring changes?

WestLangley commented:
@bhouston Sure. Merge and fix later.

mrdoob commented:
Thanks guys!

mrdoob commented:
One thing though....

Is there a way to make this the default? In the previous approach (`renderer.premultipliedAlpha`) I liked that we could just se that to `true` by default.

With the new `material.blending` approach (which I like better) seems like the user needs to change the blending by hand, it won't just work by setting `transparent: true` in the material.

bhouston commented:
Well you could set Material.blending = THREE.PremultipliedAlphaBlending in Material.js?

bhouston commented:
We could also rename THREE.PremultipliedAlphaBlending to be THREE.PremultipledAlphaNormalBlending, which suggests this is the normal mode.  It is a bit long through.  We can then have other THREE.PremultipledAlphaAddBlending, etc.

mrdoob commented:
> Well you could set Material.blending = THREE.PremultipliedAlphaBlending in Material.js?

I think that should do. However, what happens if the user wants `premultipliedAlpha` but with `AdditiveBlending`?

WestLangley commented:
Discussion continued in #8255
